### PR TITLE
miner: Ensure pending state ready for tx execution

### PR DESCRIPTION
**Pull request by piersy**
This change ensures that the pending state returned by Miner.Pending is
ready for transactions to be applied without first requiring `StateDB.Prepare` to
be called.

Prior to [EIP-2929](https://github.com/ethereum/go-ethereum/pull/21509) `StateDB.Prepare` was only required to be called on the pending state before executing a transaction if the user needed to retrieve the transaction logs. 

The changes in [EIP-2929](https://github.com/ethereum/go-ethereum/pull/21509) changed the way that gas usage was calculated for transactions, such that users were now also required to call `StateDB.Prepare` before transaction execution to ensure correct calculation of gas usage. However [EIP-2929](https://github.com/ethereum/go-ethereum/pull/21509) did not update `worker.pending` or any of the users of the pending state to ensure correct gas usage calculations.

Without this change transactions executed on the pending state that access some of the same state as the prior transaction will report incorrect gas usage.

Originally fixed here: https://github.com/celo-org/celo-blockchain/pull/1858

I didn't see any easy way to test this, but hoping someone could point out how it should be tested.

### CODE REVIEW

File: miner/worker.go [hunk 1]
```diff
@@ -343,6 +343,11 @@ func (w *worker) pending() (*types.Block, *state.StateDB) {
 	if w.snapshotState == nil {
 		return nil, nil
 	}
+	snapshotStateCopy := w.snapshotState.Copy()
+	// Call Prepare to ensure that access logs from the previously applied
+	// transaction do not affect gas use calculation for the any subsequently
+	// applied transactions.
+	snapshotStateCopy.Prepare(common.Hash{}, 0)
```
Comment (rjl493456442) [inline miner/worker.go:350]: Firstly, this `snapshotStateCopy` is not returned, it's noop.

And secondly I don't think it's a good approach to fix it. I would vote for fixing it by reset accessList here https://github.com/ethereum/go-ethereum/blob/master/core/state/statedb.go#L816

**Review by rjl493456442 — COMMENTED** [review]
I think it's a valid issue, thanks for catching it. However I am not in favour of this approach, please check the comment.

File: miner/worker.go (hunk 1 above)
Comment (piersy) [reply miner/worker.go:350]: Well spotted! My original intent was to return the `snapshotStateCopy`  :D

I wasn't convinced by this approach either, I'm definitely in favour of your suggestion, so let's see if it works!

**Review by rjl493456442 — APPROVED** [review]

**Review by holiman — COMMENTED** [review]
state_prefetcher.go:L70 does not `finalize`. 
eth/tracers/api.go:L902 does not `finalize`, and calls `Prepare` with the following comment: 
```
	// Call Prepare to clear out the statedb access list
	statedb.Prepare(txctx.TxHash, txctx.TxIndex)
```

**Comment by karalabe** [comment]
The PR is a tad dangerous as it changes the semantics throughout the entire code. Would perhaps a smaller change that just fixes DoCall work:

```diff
diff --git a/internal/ethapi/api.go b/internal/ethapi/api.go
index 366d57141..36519a00f 100644
--- a/internal/ethapi/api.go
+++ b/internal/ethapi/api.go
@@ -929,6 +929,8 @@ func DoCall(ctx context.Context, b Backend, args TransactionArgs, blockNrOrHash
 
        // Execute the message.
        gp := new(core.GasPool).AddGas(math.MaxUint64)
+
+       state.Prepare(common.Hash{}, 0) // Clears out the access lift of a "dirty" pending state
        result, err := core.ApplyMessage(evm, msg, gp)
        if err := vmError(); err != nil {
                return nil, err
```

**Comment by karalabe** [comment]
Fixed on master.

**Pull request closed without merging**